sdk/src/main/java/io/dapr/runtime/Dapr.java

```diff
@@ -64,87 +82,179 @@ public static DaprRuntime getInstance() {
         .collect(Collectors.toCollection(ArrayList::new));
   }
 
+  /**
+   * {@inheritDoc}
+   */
   @Override
   public void subscribeToTopic(String topic, TopicListener listener) {
     this.handlers.putIfAbsent(topic, new TopicHandler(listener));
   }
 
+  /**
+   * {@inheritDoc}
+   */
   @Override
   public void registerServiceMethod(String name, MethodListener listener) {
     this.handlers.putIfAbsent(name, new MethodHandler(listener));
   }
 
+  /**
+   * {@inheritDoc}
+   */
   @Override
   public Mono<byte[]> handleInvocation(String name, byte[] payload, Map<String, String> metadata) {
-    Function<HandleRequest, Mono<byte[]>> handler = this.handlers.get(name);
+    Function<Message, Mono<byte[]>> handler = this.handlers.get(name);
 
     if (handler == null) {
       return Mono.error(new DaprException("INVALID_METHOD_OR_TOPIC", "Did not find handler for : " + (name == null ? "" : name)));
     }
 
     try {
-      Map<String, Object> map = parse(payload);
-      String messageId = map.getOrDefault("id", "").toString();
-      String dataType = map.getOrDefault("datacontenttype", "").toString();
-      byte[] data = this.serializer.deserialize(map.get("data"), byte[].class);
-      return handler.apply(new HandleRequest(messageId, dataType, data, metadata)).switchIfEmpty(EMPTY_BYTES_ASYNC);
+      Message message = this.serializer.deserialize(payload, Message.class);

```

Comment:
Does this only work for topics (pub/sub) messages?  It looks like it's used for both subscriptions and method invocations from the way registrations are done, but this call seems to result in a later call deserializeTopicMessage() which looks like it's only for subscriptions.  Also, what format does `Message` correspond to?  It seems to resemble the CloudEventsEnvelope in:

https://github.com/dapr/components-contrib/blob/ede980c7ff320bc38f3284d12793910f4f4f4d93/pubsub/envelope.go

but it doesn't match exactly.

Reply:
Sorry this was posted on the wrong PR.